### DISCUSSION TITLE
Build EdgeTX on M1 Pro MAC : arm-none-eabi-gcc error

### DISCUSSION THREAD

**Discussion by Batambatam** [EdgeTX/edgetx#1627] [Q&A]
Hi,

I am trying to build EdgeTX on a M1 Pro Mac and followed the instructions on this page : 
https://github.com/EdgeTX/edgetx/wiki/Build-instructions-using-macOS-10.15-%28Catalina%29

Used this : `cmake -DPCB=X7 -DPCBREV=TPRO -DCMAKE_PREFIX_PATH=$QTDIR -DGCC_ARM_PATH=/Applications/ARM/bin/ -DPYTHON_EXECUTABLE=/usr/local/bin/python3 ..`

But I keep having this error when `make firmware` :

```
[ 16%] Building C object radio/src/targets/common/arm/stm32/bootloader/CMakeFiles/bootloader.dir/__/f2/system_stm32f2xx.c.o
arm-none-eabi-gcc: error: x86_64: No such file or directory
arm-none-eabi-gcc: error: unrecognized command-line option '-arch'; did you mean '-march='?
make[3]: *** [radio/src/targets/common/arm/stm32/bootloader/CMakeFiles/bootloader.dir/__/f2/system_stm32f2xx.c.o] Error 1
make[2]: *** [radio/src/targets/common/arm/stm32/bootloader/CMakeFiles/bootloader.dir/all] Error 2
make[1]: *** [radio/src/CMakeFiles/firmware.dir/rule] Error 2
make: *** [firmware] Error 2
```

I do have the right CMakeLists.txt with CMAKE_HOST_APPLE
Tried setting an env variable `CMAKE_HOST_APPLE=1` but no luck either.
Any help would be appreciated...
**Thanks !**

**Comment by stephenaa** [comment]
Try to set  CMAKE_APPLE_SILICON_PROCESSOR to 'arm' before your build:

export CMAKE_APPLE_SILICON_PROCESSOR=arm

**Reply by Batambatam** [reply]
Thanks, I tried and I still get the same error.

Here are my env variables :

```
declare -x APPLE="1"
declare -x CMAKE_APPLE_SILICON_PROCESSOR="arm"
declare -x HOME="/Users/batam"
declare -x LANG="fr_FR.UTF-8"
declare -x LOGNAME="batam"
declare -x LaunchInstanceID="8B26B9E0-6288-46C1-BCA8-7D7D71430EAD"
declare -x MAKE_HOST_APPLE="1"
declare -x OLDPWD="/Users/batam/edgetx"
declare -x PATH="/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Applications/VMware Fusion.app/Contents/Public"
declare -x PWD="/Users/batam/edgetx/build"
declare -x QTDIR="/Users/batam/5.12.9/clang_64"
declare -x QT_PLUGIN_PATH="/Users/batam/5.12.9/clang_64/plugins"
declare -x SECURITYSESSIONID="186ac"
declare -x SHELL="/bin/bash"
declare -x SHLVL="1"
declare -x SSH_AUTH_SOCK="/private/tmp/com.apple.launchd.ixVKb598hx/Listeners"
declare -x TERM="xterm-256color"
declare -x TERM_PROGRAM="Apple_Terminal"
declare -x TERM_PROGRAM_VERSION="443"
declare -x TERM_SESSION_ID="4DE922A3-CB1B-43EA-A4F8-3565E9F3A3C8"
declare -x TMPDIR="/var/folders/rb/d5fmzxwx56z1thvxbnzh66dh0000gn/T/"
declare -x USER="batam"
declare -x XPC_FLAGS="0x0"
declare -x XPC_SERVICE_NAME="0"
declare -x __CFBundleIdentifier="com.apple.Terminal"

```

**Comment by stephenaa** [comment]
Are you building main or the 2.6 branch? I can build main on my M1 with this:
```
export QTDIR=/opt/homebrew/Cellar/qt@5/5.15.2_1
export CMAKE_APPLE_SILICON_PROCESSOR=arm
cmake -DPCB=X10 -DPCBREV=TX16S  -DCMAKE_PREFIX_PATH=$QTDIR  -DGCC_ARM_PATH=/Applications/ARM/bin/  ..

```
(and then a "make firmware" or which target I want to build)

**Comment by Batambatam** [comment]
I am trying to build main branch (2.7.0)
Damned, I used your cmake command and still getting the error.
I don't have the same version of QT than you, but I guess this is not a problem here.

I may start over and try again the whole procedure... I don't know.

**Comment by stephenaa** [comment]
Ahh, I have also added the ARM toolchain to my path: "/Applications/ARM/bin"

Try that :-)

**Reply by Batambatam** [reply]
OK I restarted from the beginning of the guide, now it's compiling the firmware
I think I screwed the installation of homebrew at the first try.

Thanks, I really appreciate that you take time to help me 👍 

Now I try to build companion : everything compile then I have a linking issue.
CMAKE_APPLE_SILICON_PROCESSOR=arm is set.

I don't understand why it keeps compiling for x86_64...

```
ld: warning: ignoring file /Users/batam/5.12.9/clang_64/lib/QtCore.framework/QtCore, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
Undefined symbols for architecture arm64:
  "_SDL_GetError", referenced from:
      _main in companion.cpp.o
  "_SDL_Init", referenced from:
      _main in companion.cpp.o
  "_SDL_JoystickClose", referenced from:
      Joystick::~Joystick() in libsimulation.a(joystick.cpp.o)
      Joystick::close() in libsimulation.a(joystick.cpp.o)
      Joystick::open(int) in libsimulation.a(joystick.cpp.o)
  "_SDL_JoystickGetAxis", referenced from:
      Joystick::processEvents() in libsimulation.a(joystick.cpp.o)
      Joystick::getAxisValue(int) in libsimulation.a(joystick.cpp.o)
  "_SDL_JoystickGetBall", referenced from:
      Joystick::processEvents() in libsimulation.a(joystick.cpp.o)
  "_SDL_JoystickGetButton", referenced from:
      Joystick::processEvents() in libsimulation.a(joystick.cpp.o)
  "_SDL_JoystickGetHat", referenced from:
      Joystick::processEvents() in libsimulation.a(joystick.cpp.o)
  "_SDL_JoystickName", referenced from:
      Joystick::Joystick(QObject*, int, bool, int) in libsimulation.a(joystick.cpp.o)
  "_SDL_JoystickNumAxes", referenced from:
      Joystick::open(int) in libsimulation.a(joystick.cpp.o)
  "_SDL_JoystickNumBalls", referenced from:
      Joystick::open(int) in libsimulation.a(joystick.cpp.o)
  "_SDL_JoystickNumButtons", referenced from:

[...]

ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [companion.app/Contents/MacOS/companion] Error 1
make[2]: *** [companion/src/CMakeFiles/companion.dir/all] Error 2
make[1]: *** [companion/src/CMakeFiles/companion.dir/rule] Error 2
make: *** [companion] Error 2
```

**Reply by stephenaa** [reply]
Did you run cmake again in a clean directory?
I got it to work after some effort, but in the end it ended up being quite little that needed to be changed - and that should be on the 2.7 branch now.

**Reply by Batambatam** [reply]
I did a make clean and run cmake again.
I saw your PR was merged and you are several dev people that do not have issues since.

As I am still learning all this stuff of compiling on an arm Mac, it may be the reason why I am struggling to get it work.

Error seems to be about QT libraries, and 5.xx do not support Apple Silicon. Could it be the reason (I have Qt 5.12.9) ?
Or how do I compile for x86_64 target from my apple silicon ?

**Comment by Batambatam** [comment]
OK, I managed to build Companion and simulator.
I first built Qt5.15.3 libs for Apple Silicon.
I followed instructions on this page : https://eric-pierce.com/compiling-nextcloud-for-apple-silicon/
Then I also had to copy my SDL 1.2.15 libs (brew arm64-monterey version) to /usr/local/lib/ 

It's running native on my M1 now 👍

**Comment by stephenaa** [comment]
Good to hear that. They are nice machines, and pretty good development rigs 👍

Sendt fra min iPhone
